### PR TITLE
Match wild err arm improvements

### LINKED ISSUE #5024: `match_wild_err_arm` triggers weirdly
If you match on `Err(_)` then it triggers the `match_wild_err_arm` lint and clippy tells you to match on each error individually, but if you match on `Err(_e)` it doesn't trigger at all, even though that's ignoring all possible errors just the same.

Code that brought this up
```rust
  #[inline]
  #[must_use]
  pub fn from_array_len(data: A, len: usize) -> Self {
    match Self::try_from_array_len(data, len) {
      Ok(out) => out,
      Err(_) => {
        panic!("ArrayishVec: length {} exceeds capacity {}!", len, A::CAPACITY)
      }
    }
  }
```

* This should probably only trigger on enum errors, not struct errors.
* Using a name other than `_` will still match on everything, which is what this lint theoretically doesn't link, so alternate names should also still trigger the lint

### PR DESCRIPTION
**Pull request by ThibsG**
This lint should trigger on other identifiers which have `_` prefix (such as `_e`) and only if they are unused in the panic block.

_Note_: the `is_unused` function is greatly inspired from `pat_is_wild` function in [loops lints](https://github.com/rust-lang/rust-clippy/blob/43ac9416d935942d6c7d2b2e0c876c551652c4ec/clippy_lints/src/loops.rs#L1689).
I've been considering doing some refactoring, maybe in utils. Maybe this PR or a new one. What do you think ?

fixes #5024

changelog: none

### CODE REVIEW

File: clippy_lints/src/matches.rs [hunk 1]
```diff
@@ -468,26 +471,70 @@ fn is_wild<'tcx>(pat: &impl std::ops::Deref<Target = Pat<'tcx>>) -> bool {
...
+                        if is_panic_block(block);
+                        then {
+                            // `Err(_)` or `Err(_e)` arm with `panic!` found
+                            span_note_and_lint(cx,
+                                            MATCH_WILD_ERR_ARM,
+                                            arm.pat.span,
+                                            &format!("`Err({})` will match all errors, maybe not a good idea", &ident_bind_name),
+                                            arm.pat.span,
+                                            "to remove this warning, match each error separately \
+                                                or use `unreachable!` macro");
```
Comment (flip1995) [inline clippy_lints/src/matches.rs]: Can you format this as

```rust
span_note_and_lint(
    cx,
    MATCH_WILD_ERR_ARM,
    ...,
);
```

File: clippy_lints/src/matches.rs [hunk 2]
```diff
@@ -468,26 +471,70 @@ fn is_wild<'tcx>(pat: &impl std::ops::Deref<Target = Pat<'tcx>>) -> bool {
...
+                    if_chain! {
+                        if matching_wild;
+                        if let ExprKind::Block(ref block, _) = arm.body.kind;
+                        if is_panic_block(block);
+                        then {
+                            // `Err(_)` or `Err(_e)` arm with `panic!` found
+                            span_note_and_lint(cx,
+                                            MATCH_WILD_ERR_ARM,
+                                            arm.pat.span,
+                                            &format!("`Err({})` will match all errors, maybe not a good idea", &ident_bind_name),
```
Comment (flip1995) [inline clippy_lints/src/matches.rs]:
```suggestion
                                            &format!("`Err({})` matches all errors", &ident_bind_name),
```

File: clippy_lints/src/matches.rs (hunk 1 above)
Comment (flip1995) [inline clippy_lints/src/matches.rs]:
```suggestion
                                            "match each error separately or use the error output",
```

Comment (ThibsG) [reply clippy_lints/src/matches.rs]: Sure, my fmt seems broken :face_with_head_bandage:

Comment (flip1995) [reply clippy_lints/src/matches.rs]: Nah, not your fault. Rustfmt doesn't run in (most) macros. Also you didn't touch this code, so double not your fault 😄 But while your at it... 😉

**Comment by flip1995** [comment]
Thanks!

@bors r+

**Pull request merged**